From 35685e06752e4716590675c2a09b988e5b565adc Mon Sep 17 00:00:00 2001
From: Abhinaba Rakshit <abhinaba.rakshit@oss.qualcomm.com>
Date: Fri, 21 Nov 2025 14:46:29 +0530
Subject: [PATCH 3/3] FROMLIST: soc: qcom: ice: Set ICE clk to TURBO on probe

MMC controller lacks a clock scaling mechanism, unlike the UFS
controller. By default, the MMC controller is set to TURBO mode
during probe, but the ICE clock remains at XO frequency,
leading to read/write performance degradation on eMMC.

To address this, set the ICE clock to TURBO during probe to
align it with the controller clock. This ensures consistent
performance and avoids mismatches between the controller
and ICE clock frequencies.

For platforms where ICE is represented as a separate device,
use the OPP framework to vote for TURBO mode, maintaining
proper voltage and power domain constraints. For legacy
targets where ICE is integrated with the storage controller,
fall back to using standard clock APIs to set the frequency.

Signed-off-by: Abhinaba Rakshit <abhinaba.rakshit@oss.qualcomm.com>
Upstream-Status: Submitted
---
 drivers/soc/qcom/ice.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/soc/qcom/ice.c b/drivers/soc/qcom/ice.c
index 6d39a52a811f..88b233a3833b 100644
--- a/drivers/soc/qcom/ice.c
+++ b/drivers/soc/qcom/ice.c
@@ -604,6 +604,11 @@ static struct qcom_ice *qcom_ice_create(struct device *dev,
 		if (clk_index < 0)
 			return ERR_PTR(clk_index);
 
+		/* Vote for maximum clock rate for maximum performance */
+		err = clk_set_rate(engine->core_clk, INT_MAX);
+		if (err)
+			dev_warn(dev, "Failed boosting the ICE clk to TURBO\n");
+
 		break;
 	}
 
@@ -624,6 +629,11 @@ static struct qcom_ice *qcom_ice_create(struct device *dev,
 		}
 		engine->has_opp = (err == 0);
 
+		/* Vote for maximum clock rate for maximum performance */
+		err = dev_pm_opp_set_rate(dev, INT_MAX);
+		if (err)
+			dev_warn(dev, "Failed boosting the ICE clk to TURBO\n");
+
 		/* Since, there is only one clock
 		 * index can be set as 0
 		 */
-- 
2.34.1

